### ISSUE TITLE
OAUTH_DEBUG=true causes StackLocked exception

### ISSUE THREAD

**Issue by miyagawa** [oauth-xx/oauth2#189]
When you call `access_token.get` multiple times on the same AccessToken object where `OAUTH_DEBUG` environment is set to true, you'll get an exception: "can't modify middleware stack after making a request (Faraday::RackBuilder::StackLocked)"

Confirmed with faraday 0.9.0 and oauth2 gem 1.0.0.

Run the following script with or without `OAUTH_DEBUG=true`:

``` ruby
require 'oauth2'

client = OAuth2::Client.new('id', 'secret', site: 'http://example.com')
token = OAuth2::AccessToken.new(client, 'foo')

token.get "/"
token.get "/"
```

Without the environment, it runs just fine with no errors. With the debug environment set:

```
> OAUTH_DEBUG=true be ruby ~/tmp/oauth2_access_token_get.rb
I, [2014-08-18T19:50:49.005699 #32581]  INFO -- : get http://example.com/
...
connection: "close"
/Users/miyagawa/.rbenv/versions/2.1.1/lib/ruby/gems/2.1.0/gems/faraday-0.9.0/lib/faraday/rack_builder.rb:199:in `raise_if_locked': can't modify middleware stack after making a request (Faraday::RackBuilder::StackLocked)
        from /Users/miyagawa/.rbenv/versions/2.1.1/lib/ruby/gems/2.1.0/gems/faraday-0.9.0/lib/faraday/rack_builder.rb:86:in `use'
```

**Comment by mlwilkerson** [comment]
I'm seeing the same thing with faraday 0.9.1 and oauth2 gem.

**Comment by djberg96** [comment]
I see the same thing with faraday 0.9.2 and oauth2 version 1.2.0.

**Comment by james-prickett** [comment]
I have this problem with ruby 2.2.6, oauth2 (1.3.0), and faraday (0.10.1).   My solution was to move the line below, found at client.rb:95 to client.rb:63 at the end of the block that initializes the connection object. 

connection.response :logger, ::Logger.new($stdout) if ENV['OAUTH_DEBUG'] == 'true'

**Comment by joevandyk** [comment]
i ran into this as well

**Comment by JasonLunn** [comment]
Also encountered this today, and locally developed a very similar workaround to @gafrom's PR #294

**Comment by kevinelliott** [comment]
I also encountered this today.

**Comment by josephpage** [comment]
Fixed by #294 and #142
Will be shipped in the next release.

**Comment by fdoxyz** [comment]
When using `omniauth-facebook` we had enabled `OAUTH_DEBUG=true` and this same error was raised. Faraday retried the POST request after the first failure but Facebook responds with this:

```
(facebook) Authentication failure! invalid_credentials: OAuth2::Error, {"message"=>"This authorization code has been used.", "type"=>"OAuthException", "code"=>100, "error_subcode"=>36009, "fbtrace_id"=>"AfRdd7duq212Vsx"}
```

<img width="1211" alt="Screen Shot 2020-09-23 at 10 32 05" src="https://user-images.githubusercontent.com/6045239/94049414-34234280-fd92-11ea-938b-6da623caa4fe.png">

The gem versions we're using:

```
omniauth (1.9.1)
oauth2 (1.4.4)
omniauth-oauth2 (1.7.0)
omniauth-facebook (7.0.0)
faraday (1.0.1)
```

We do have other custom middleware (as the backtrace shows from the logs) but removing the `OAUTH_DEBUG=true` solved our problem.

**Comment by suupic** [comment]
same problem：

```
Authentication failure! can't modify middleware stack after making a request: Faraday::RackBuilder::StackLocked, can't modify middleware stack after making a request

Faraday::RackBuilder::StackLocked (can't modify middleware stack after making a request):
```

The gem versions we're using:
```
faraday (1.3.0)
faraday-net_http (1.0.1)
omniauth (2.0.3)
omniauth-oauth2 (1.7.1)
```

**Comment by IlkhamGaysin** [comment]
+1 I have the same error

**Comment by alexanderadam** [comment]
I'm also seeing this with these dependencies:

```
oauth2 (1.4.7)
  faraday (>= 0.8, < 2.0)
  jwt (>= 1.0, < 3.0)
  multi_json (~> 1.3)
  multi_xml (~> 0.5)
  rack (>= 1.2, < 3)
```

**Comment by pboling** [comment]
Thanks for the reports.  I'll work on a fix.

**Comment by pboling** [comment]
Stack issues were caused by a recursive dependency between response.error and error.response.  This is fixed in just released version 1.4.9!
